compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -486,6 +532,133 @@ abstract class GenJSCode extends plugins.PluginComponent
       classDefinition
     }
 
+    /** Generate an instance of an anonymous Scala.js defined class inline
+     *
+     *  @param sym Class to generate the instance of
+     *  @param args Arguments to the constructor
+     *  @param pos Position of the original New tree
+     */
+    def genAnonSJSDefinedNew(sym: Symbol, args: List[js.Tree],
+        pos: Position): js.Tree = {
+      assert(isScalaJSDefinedAnonJSClass(sym),
+          "Generating AnonSJSDefinedNew of non anonymous SJSDefined JS class")
+
+      // Find the ClassDef for this anonymous class
+      val classDef = consumeLazilyGeneratedAnonClass(sym)
+
+      // Generate a normal SJSDefinedJSClass
+      val origJsClass =
+        nestedGenerateClass(sym)(genScalaJSDefinedJSClass(classDef))
+
+      // Partition members into the ones we inline and the ones we keep
+      val retainedMembers = ListBuffer.empty[js.Tree]
+      var constructor: Option[js.MethodDef] = None
+      val exportedMembers = ListBuffer.empty[js.Tree]
+
+      origJsClass.defs.foreach {
+        case _: js.FieldDef => // ignore. we get it from the ctor
+
+        case mdef: js.MethodDef =>
+          mdef.name match {
+            case _: js.Ident =>
+              assert(mdef.static, "Non-static method in SJS defined JS class")
+              retainedMembers += mdef
+
+            case js.StringLiteral(name) =>
+              if (name == "constructor") {
+                assert(constructor.isEmpty, "two ctors in class")
+                constructor = Some(mdef)
+              } else {
+                exportedMembers += mdef
+              }
+          }
+
+        case property: js.PropertyDef =>
+          exportedMembers += property
+
+        case tree =>
+          sys.error("Unexpected tree: " + tree)
+      }
+
+      // Only emit static members to class
+      generatedClasses += sym -> origJsClass.copy(
+          defs = retainedMembers.toList)(
+          origJsClass.optimizerHints)(origJsClass.pos)
+
+      // Construct inline class definition
+      assert(constructor.isDefined, "No ctor found")
+      val js.MethodDef(_, _, ctorParams, _, ctorBody) = constructor.get
+
+      assert(ctorParams.size == args.size)
+      val argDefs = for ((param, arg) <- ctorParams zip args)
+        yield js.VarDef(param.name, param.ptpe, param.mutable, arg)(param.pos)
```

Comment:
Such a rewriting would also protect from applying the transform twice on a nested anonymous JS class (where the constructor of the outer one defines an anonymous inner one!) -> probably needs a test, that one ^^